pkg/controller/controller.go

```diff
@@ -177,96 +194,163 @@ func (tc *TenantController) processItem(e Event) error {
 	// process events based on its type
 	switch e.eventType {
 	case Create:
-		return tc.syncResources(e.tenantId)
+		return tc.onTenantCreate(e.tenantId)
+	case Update:
+		return tc.onTenantUpdate(e.tenantId)
 	case Delete:
 		klog.Warningf("deleting tenant %v", e.tenantId)
-		return tc.deleteResources(e.tenantId)
+		return nil
+	}
+	return nil
+}
+
+// onTenantCreate handles the CREATE event of a Tenant.
+func (tc *TenantController) onTenantCreate(tenantId string) error {
+	tenant, err := tc.tenantClient.Tenants().Get(context.TODO(), tenantId, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if tenant.ObjectMeta.DeletionTimestamp.IsZero() && !util.ContainString(tenant.ObjectMeta.Finalizers, tenantFinalizerKey) {
+		tenant.ObjectMeta.Finalizers = append(tenant.ObjectMeta.Finalizers, tenantFinalizerKey)
+		if _, err := tc.tenantClient.Tenants().Update(context.TODO(), tenant, metav1.UpdateOptions{}); err != nil {
+			return err
+		}
+	}
+
+	return tc.syncResources(tenantId)
```

Comment:
I don't think there is a consistency issue because syncResources is idempotent. That is, it first gets/lists the resources, then compares the current state to the desired state and decides what to do.

Also, onTenantCreate has been deprecated and replaced by onTenantCreateOrUpdate. I forgot to delete it. I have removed it in the new commit.

Reply:
So "create" actually means "added to informer store", not "create tenant".  Would it be better to continue to use the name "add" (as is consistent with `ResourceEventHandlerFuncs`)?

Reply:
Yes, I think this is a very good suggestion. I changed the name in the latest commit.